File: dp/c/data_mapper.c
// Data Mapper Pattern in C

#include <stdio.h>
#include <string.h>

typedef struct { int id; char name[32]; char email[64]; } User;
typedef struct { int user_id; char user_name[32]; char user_email[64]; char created_at[20]; } UserRow;

UserRow user_to_row(const User* u) {
    UserRow r;
    r.user_id = u->id;
    snprintf(r.user_name, 32, "%s", u->name);
    snprintf(r.user_email, 64, "%s", u->email);
    snprintf(r.created_at, 20, "2024-01-01");
    return r;
}

User row_to_user(const UserRow* r) {
    User u;
    u.id = r->user_id;
    snprintf(u.name, 32, "%s", r->user_name);
    snprintf(u.email, 64, "%s", r->user_email);
    return u;
}

#define MAX_ROWS 16
typedef struct { UserRow rows[MAX_ROWS]; int count; } Database;

void db_save(Database* db, const User* u) {
    UserRow r = user_to_row(u);
    printf("[DB] Save: id=%d name=%s\n", r.user_id, r.user_name);
    db->rows[db->count++] = r;
}

User* db_find(Database* db, int id, User* out) {
    for (int i = 0; i < db->count; i++)
        if (db->rows[i].user_id == id) { *out = row_to_user(&db->rows[i]); return out; }
    return NULL;
}

int main(void) {
    Database db = {.count = 0};
    User alice = {1, "Alice", "alice@example.com"};
    User bob = {2, "Bob", "bob@example.com"};
    db_save(&db, &alice);
    db_save(&db, &bob);

    User found;
    if (db_find(&db, 1, &found))
        printf("Found: id=%d name=%s email=%s\n", found.id, found.name, found.email);
    return 0;
}

File: dp/rust/src/data_mapper.rs
// Data Mapper Pattern in Rust
// Maps between domain objects and persistence representation

use std::collections::HashMap;

// Domain model (clean, no persistence concerns)
#[derive(Debug, Clone)]
struct User {
    id: u32,
    name: String,
    email: String,
}

// Persistence representation (row in database)
#[derive(Debug)]
struct UserRow {
    user_id: u32,
    user_name: String,
    user_email: String,
    created_at: String,
}

// Data Mapper: translates between domain and persistence
struct UserMapper;

impl UserMapper {
    fn to_domain(row: &UserRow) -> User {
        User {
            id: row.user_id,
            name: row.user_name.clone(),
            email: row.user_email.clone(),
        }
    }

    fn to_row(user: &User) -> UserRow {
        UserRow {
            user_id: user.id,
            user_name: user.name.clone(),
            user_email: user.email.clone(),
            created_at: "2024-01-01".to_string(),
        }
    }
}

// Simulated database
struct Database {
    rows: HashMap<u32, UserRow>,
}

impl Database {
    fn new() -> Self {
        Database { rows: HashMap::new() }
    }

    fn save(&mut self, user: &User) {
        let row = UserMapper::to_row(user);
        println!("[DB] Saving row: {:?}", row);
        self.rows.insert(row.user_id, row);
    }

    fn find(&self, id: u32) -> Option<User> {
        self.rows.get(&id).map(UserMapper::to_domain)
    }

    fn find_all(&self) -> Vec<User> {
        self.rows.values().map(UserMapper::to_domain).collect()
    }
}

fn main() {
    let mut db = Database::new();

    // Domain objects
    let alice = User { id: 1, name: "Alice".into(), email: "alice@example.com".into() };
    let bob = User { id: 2, name: "Bob".into(), email: "bob@example.com".into() };

    // Save through mapper
    db.save(&alice);
    db.save(&bob);

    // Load through mapper
    println!("\nFind by id=1: {:?}", db.find(1));
    println!("All users: {:?}", db.find_all());
}
